File: dexfile/src/main/java/com/github/netomi/bat/dexfile/MapItem.java
/*
 *  Copyright (c) 2020 Thomas Neidhart.
 *
 *  Licensed under the Apache License, Version 2.0 (the "License");
 *  you may not use this file except in compliance with the License.
 *  You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 */
package com.github.netomi.bat.dexfile;

import com.github.netomi.bat.dexfile.io.DexDataInput;
import com.github.netomi.bat.dexfile.io.DexDataOutput;

import java.util.Objects;

/**
 * A class representing a map item inside a dex file.
 * <p>
 * Note: this class is not intended for general use, its automatically created when writing dex files.
 *
 * @see <a href="https://source.android.com/devices/tech/dalvik/dex-format#map-item">map item @ dex format</a>
 *
 * @author Thomas Neidhart
 */
public class MapItem
extends      DexContent
{
    // member fields are package private for convenience reasons.

    int type;     // ushort
    //int unused; // ushort
    int size;     // uint
    int offset;   // uint

    static MapItem of(int type) {
        return new MapItem(type);
    }

    static MapItem of(int type, int size) {
        return new MapItem(type, size);
    }

    static MapItem readContent(DexDataInput input) {
        MapItem mapItem = new MapItem();
        mapItem.read(input);
        return mapItem;
    }

    private MapItem() {}

    private MapItem(int type) {
        this(type, 0);
    }

    private MapItem(int type, int size) {
        this.type = type;
        this.size = size;
    }

    public int getType() {
        return type;
    }

    public int getSize() {
        return size;
    }

    public int getOffset() {
        return offset;
    }

    @Override
    protected void read(DexDataInput input) {
        type   = input.readUnsignedShort();
        input.readUnsignedShort();
        size   = input.readInt();
        offset = input.readInt();
    }

    @Override
    protected void write(DexDataOutput output) {
        output.writeUnsignedShort(type);
        output.writeUnsignedShort(0x0);
        output.writeInt(size);
        output.writeInt(offset);
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;
        MapItem other = (MapItem) o;
        return type == other.type &&
               size == other.size;
    }

    @Override
    public int hashCode() {
        return Objects.hash(type, size);
    }

    @Override
    public String toString() {
        return String.format("MapItem[type=0x%04x,size=%5d,offset=%7d]", type, size, offset);
    }
}

File: dexfile/src/main/java/com/github/netomi/bat/dexfile/MapItem.kt
/*
 *  Copyright (c) 2020 Thomas Neidhart.
 *
 *  Licensed under the Apache License, Version 2.0 (the "License");
 *  you may not use this file except in compliance with the License.
 *  You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 */
package com.github.netomi.bat.dexfile

import com.github.netomi.bat.dexfile.io.DexDataInput
import com.github.netomi.bat.dexfile.io.DexDataOutput
import com.google.common.base.Preconditions
import java.util.*

/**
 * A class representing a map item inside a dex file.
 *
 * Note: this class is not intended for general use, its automatically created when writing dex files.
 *
 * @see [map item @ dex format](https://source.android.com/devices/tech/dalvik/dex-format.map-item)
 */
class MapItem private constructor(_type: Int = 0, _size: Int = 0, _offset: Int = 0): DexContent() {

    var type: Int = _type
        private set

    var size: Int = _size
        internal set(value) {
            Preconditions.checkArgument(value >= 0, "size must not be negative")
            field = value
        }

    var offset: Int = _offset
        internal set(value) {
            Preconditions.checkArgument(value >= 0, "offset must not be negative")
            field = value
        }

    override fun read(input: DexDataInput) {
        type = input.readUnsignedShort()
        input.readUnsignedShort()
        size = input.readInt()
        offset = input.readInt()
    }

    override fun write(output: DexDataOutput) {
        output.writeUnsignedShort(type)
        output.writeUnsignedShort(0x0)
        output.writeInt(size)
        output.writeInt(offset)
    }

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (other == null || javaClass != other.javaClass) return false
        val o = other as MapItem
        return type   == o.type &&
               size   == o.size &&
               offset == o.offset
    }

    override fun hashCode(): Int {
        return Objects.hash(type, size)
    }

    override fun toString(): String {
        return "MapItem[type=0x%04x,size=%5d,offset=%7d]".format(type, size, offset)
    }

    companion object {
        fun of(type: Int, size: Int = 0): MapItem {
            Preconditions.checkArgument(size >= 0, "size must not be negative")
            return MapItem(type, size)
        }

        fun readContent(input: DexDataInput): MapItem {
            val mapItem = MapItem()
            mapItem.read(input)
            return mapItem
        }
    }
}